src/CameraOnvif.cpp

```diff
@@ -0,0 +1,408 @@
+    SetImagingSettings.ImagingSettings->ColorSaturation = &new_color_saturation;
+    _timg__SetImagingSettingsResponse SetImagingSettingsResponse;
+    setCredentials();
+    if (m_private->proxy_image->SetImagingSettings(&SetImagingSettings, SetImagingSettingsResponse))
+        reportError();
+}
+
+void CameraOnvif::setContrast(float percentage){
+    float new_contrast = (contrast[1] - contrast[0]) * percentage + contrast[0];
+    _timg__SetImagingSettings SetImagingSettings;
+    SetImagingSettings.VideoSourceToken = video_token;
+    SetImagingSettings.ImagingSettings = new tt__ImagingSettings20();
+    SetImagingSettings.ImagingSettings->Contrast = &new_contrast;
+    _timg__SetImagingSettingsResponse SetImagingSettingsResponse;
+    setCredentials();
+    if (m_private->proxy_image->SetImagingSettings(&SetImagingSettings, SetImagingSettingsResponse))
+        reportError();
+}
+
+void CameraOnvif::setResolution(int width, int height){
+    bool valid = true;
+    for (auto & e : m_private->video_options->Options->H264->ResolutionsAvailable){
+        if (e->Width == width && e->Height == height){
+            valid = true;
+        }
+    }
+
+    _trt__GetVideoEncoderConfiguration GetVideoEncoderConfiguration;
+    _trt__GetVideoEncoderConfigurationResponse GetVideoEncoderConfigurationResponse;
+    GetVideoEncoderConfiguration.ConfigurationToken = video_conf_token;
+
+    setCredentials();
+    if (m_private->proxy_media->GetVideoEncoderConfiguration(&GetVideoEncoderConfiguration,
+        GetVideoEncoderConfigurationResponse)){
+        reportError();
+    }
+
+    _trt__SetVideoEncoderConfiguration SetVideoEncoderConfiguration;
+    _trt__SetVideoEncoderConfigurationResponse SetVideoEncoderConfigurationResponse;
+    SetVideoEncoderConfiguration.Configuration =
+    GetVideoEncoderConfigurationResponse.Configuration;
+    // SetVideoEncoderConfiguration.Configuration->Encoding = tt__VideoEncoding__H264;
+    // SetVideoEncoderConfiguration.Configuration->Quality = 6;
+    // SetVideoEncoderConfiguration.Configuration->Multicast = new tt__MulticastConfiguration();
+    // SetVideoEncoderConfiguration.ForcePersistence = true;
+    // SetVideoEncoderConfiguration.Configuration->SessionTimeout = "10";
+    // SetVideoEncoderConfiguration.Configuration->H264 = new tt__H264Configuration();
+    // SetVideoEncoderConfiguration.soap = m_private->soap;
+    // SetVideoEncoderConfiguration.Configuration->H264->soap = m_private->soap;
+
+    // if (width == 1920 and height == 1080){
+    //     SetVideoEncoderConfiguration.Configuration->H264->H264Profile =
+    //     tt__H264Profile__High;
+    // }else if (width == 1280 and height == 960){
+    //     SetVideoEncoderConfiguration.Configuration->H264->H264Profile =
+    //     tt__H264Profile__Main;
+    // }else if (width == 1280 and height == 720){
+    //     SetVideoEncoderConfiguration.Configuration->H264->H264Profile =
+    //     tt__H264Profile__Baseline;
+    // }
+    if(!valid) {
```

Comment:
Why do you error here ? Why not just after the validation, very early in the method ?

Reply:
I had to move the verification to very early due to memory problems but I forgot to move this "throwing" too.